src/main/scala/extracells/part/PartFluidConversionMonitor.java

```diff
@@ -33,90 +32,42 @@ public boolean onActivate(EntityPlayer player, Vec3 pos) {
-							}
-							ItemStack s3 = s.copy();
-							s3.stackSize--;
-							if (s3.stackSize <= 0)
-								player.inventory.setInventorySlotContents(player.inventory.currentItem, null);
-							else
-								player.inventory.setInventorySlotContents(player.inventory.currentItem, s3);
-						}
-					}
 				}
-				return true;
-			} else if (FluidUtil.isEmpty(s2)) {
+				ItemStack simulation = FluidUtil.drainItemIntoAe(container, mon, Actionable.SIMULATE, src);
+				if (simulation == null) {
+					return true;
+				}
+				result = FluidUtil.drainItemIntoAe(container, mon, Actionable.MODULATE, src);
+			} else if (FluidUtil.isEmpty(container)) {
 				if (this.fluid == null)
 					return true;
-				IAEFluidStack extract;
-				if (s2.getItem() instanceof IFluidContainerItem) {
-					extract = mon.extractItems(FluidUtil.createAEFluidStack(
-							this.fluid, ((IFluidContainerItem) s2.getItem())
-									.getCapacity(s2)), Actionable.SIMULATE,
-							new MachineSource(this));
-				} else
-					extract = mon.extractItems(
-							FluidUtil.createAEFluidStack(this.fluid),
-							Actionable.SIMULATE, new MachineSource(this));
-				if (extract != null) {
-					if (extract.getStackSize() <= 0)
-						return true;
-					extract = mon.extractItems(extract, Actionable.MODULATE, new MachineSource(this));
-					if (extract == null || extract.getStackSize() <= 0)
-						return true;
-
-					MutablePair<Integer, ItemStack> empty1 = FluidUtil
-							.fillStack(s2, extract.getFluidStack());
-					if (empty1.left == 0) {
-						mon.injectItems(extract, Actionable.MODULATE, new MachineSource(this));
-						return true;
-					}
-					ItemStack empty = empty1.right;
-					if (empty != null) {
-						dropItems(getHost().getTile().getWorldObj(), getHost()
-								.getTile().xCoord + getSide().offsetX,
-								getHost().getTile().yCoord + getSide().offsetY,
-								getHost().getTile().zCoord + getSide().offsetZ,
-								empty);
-					}
-					ItemStack s3 = s.copy();
-					s3.stackSize = s3.stackSize - 1;
-					if (s3.stackSize == 0) {
-						player.inventory.setInventorySlotContents(
-								player.inventory.currentItem, null);
-					} else {
-						player.inventory.setInventorySlotContents(
-								player.inventory.currentItem, s3);
-					}
+				MutablePair<ItemStack, FluidStack> simulation = FluidUtil.fillItemFromAe(container, new FluidStack(this.fluid, Integer.MAX_VALUE), mon, Actionable.SIMULATE, src);
+				if (simulation == null || simulation.getLeft() == null) {
+					return true;
 				}
+				result = FluidUtil.fillItemFromAe(container, new FluidStack(this.fluid, Integer.MAX_VALUE), mon, Actionable.MODULATE, src).getLeft();
+			}
+			if (result == null) {
 				return true;
 			}
+			TileEntity tile = this.getHost().getTile();
+			ForgeDirection side = this.getSide();
+			this.dropItems(tile.getWorldObj(), tile.xCoord + side.offsetX, tile.yCoord + side.offsetY, tile.zCoord + side.offsetZ, result);
```

Comment:
Wouldn't it be better to insert stuff to the player inventory like AE2 does? Yes, `appeng.util.InventoryAdaptor` is not a part of the API but I think it makes no sense to restrict yourself to the API in EC2 case, after all fluid handling was integrated in AE2 in 1.12 and I doubt we will go that far.